---
title: "RNAseq_post-analysis"
author: "Jeff Lee"
date: '2022-05-03'
output: html_document
editor_options: 
  chunk_output_type: console
---

# GENE ONTOLOGY ENRICHMENT ANALYSIS

## Get Brain Transcriptome

 - ENCODE Mmus bulk RNA-seq brain adult male (https://www.encodeproject.org/files/ENCFF468CXD/) 

```{r}
library(tidyverse)

## Read in ENCODE transcriptome quant tsv file
encode_raw <- read_tsv("./RNAseq/external_data/ENCFF468CXD.tsv") %>%
  filter(str_detect(gene_id, "ENS")) %>%
  mutate(gene_id = str_replace(gene_id, "\\..*", ""))

## How many genes have TPM > 1 ? - Using lower cutoff due to smaller presence of glia in the whole brain
encode_raw %>% filter(TPM > 1) %>% nrow()

## Get a vector of gene_id with TPM > 1
Mmus_encode_brain_txome <- encode_raw %>%
  filter(TPM > 1) %>% pull(gene_id)

# * * * * * Add any glial transcripts that are not represented in ENCODE data

## Which transcripts from our summary dataframe are not represented in the ENCODE data? (TPM > 10)
avgTPM_wide <- readRDS("./RNAseq/quant_results/all-glia_avgTPM_wide.RDS")
avgTPM_tidy <- avgTPM_wide %>%
  pivot_longer(cols = 2:last_col(), names_to = "glia_types", values_to = "TPM")
glial_txome <- avgTPM_tidy %>%
  filter(TPM > 10) %>%
  pull(gene_id) %>% 
  unique()

## How many of the genes intersect between brain txome and glial txome?
intersect(Mmus_encode_brain_txome, glial_txome) %>% length() # ~2000 are missing in brain txome

## Add the two together and take the unique gene_id sets
Mmus_brain_glia_txome <- c(Mmus_encode_brain_txome, glial_txome) %>% unique()

saveRDS(Mmus_brain_glia_txome, "./RNAseq/external_data/Mmus_brain_glia_txome.RDS")

```

## Run Gene Ontology analysis

```{r}
## source TopGO script
source("./src/runTopGO.R")
Mmus_brain_glia_txome <- readRDS("./RNAseq/external_data/Mmus_brain_glia_txome.RDS")
df_annotated <- readRDS("./RNAseq/summary_table/df_annotated.RDS")

## Ontologies to test
ontologies <- c("BP", "MF", "CC") %>% set_names()
```

### Strict cutoffs 

```{r}
# * * * * *  Those present in protrusions

## RNA_in_protrusion >= 12
strict_set1_target <- df_annotated %>%
  filter(RNA_in_protrusion >= 12) %>%
  pull(gene_id)

## How many genes in this set? = 366
strict_set1_target %>% length()

## runTopGO
strict_set1 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = strict_set1_target)
}, .id = "ontology")

strict_set1_filtered <- strict_set1 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

# * * * * * Those preferentially enriched in protrusions

## enriched_in_protrusion >= 4
strict_set2_target <- df_annotated %>%
  filter(enriched_in_protrusion >= 4) %>%
  pull(gene_id)

## How many genes in this set? = 66
strict_set2_target %>% length()

## runTopGO
strict_set2 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = strict_set2_target)
}, .id = "ontology")

strict_set2_filtered <- strict_set2 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

# * * * * * Those translated in protrusions 

## translation_in_protrusion >= 4
strict_set3_target <- df_annotated %>%
  filter(translation_in_protrusion >= 4) %>%
  pull(gene_id)

## How many genes in this set? = 1599
strict_set3_target %>% length()

## runTopGO
strict_set3 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = strict_set3_target)
}, .id = "ontology")

strict_set3_filtered <- strict_set3 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

# * * * * * Those preferentially enriched & translated in protrusions 

## enriched_in_protrusion >= 4 & translation_in_protrusion >= 4
strict_set4_target <- df_annotated %>%
  filter(enriched_in_protrusion >= 4 & translation_in_protrusion >= 4) %>%
  pull(gene_id)

## How many genes in this set? = 31
strict_set4_target %>% length()

## runTopGO
strict_set4 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = strict_set4_target)
}, .id = "ontology")

strict_set4_filtered <- strict_set4 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

```

### Lenient cutoffs

```{r}
# * * * * *  Those present in protrusions

## RNA_in_protrusion >= 10
lenient_set1_target <- df_annotated %>%
  filter(RNA_in_protrusion >= 10) %>%
  pull(gene_id)

## How many genes in this set? = 2755
lenient_set1_target %>% length()

## runTopGO
lenient_set1 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = lenient_set1_target)
}, .id = "ontology")

lenient_set1_filtered <- lenient_set1 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

# * * * * * Those preferentially enriched in protrusions

## enriched_in_protrusion >= 3
lenient_set2_target <- df_annotated %>%
  filter(enriched_in_protrusion >= 3) %>%
  pull(gene_id)

## How many genes in this set? = 491
lenient_set2_target %>% length()

## runTopGO
lenient_set2 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = lenient_set2_target)
}, .id = "ontology")

lenient_set2_filtered <- lenient_set2 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

# * * * * * Those translated in protrusions 

## translation_in_protrusion >= 3
lenient_set3_target <- df_annotated %>%
  filter(translation_in_protrusion >= 3) %>%
  pull(gene_id)

## How many genes in this set? = 3656
lenient_set3_target %>% length()

## runTopGO
lenient_set3 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = lenient_set3_target)
}, .id = "ontology")

lenient_set3_filtered <- lenient_set3 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

# * * * * * Those preferentially enriched & translated in protrusions 

## enriched_in_protrusion >= 3 & translation_in_protrusion >= 3
lenient_set4_target <- df_annotated %>%
  filter(enriched_in_protrusion >= 3 & translation_in_protrusion >= 3) %>%
  pull(gene_id)

## How many genes in this set? = 278
lenient_set4_target %>% length()

## runTopGO
lenient_set4 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = lenient_set4_target)
}, .id = "ontology")

lenient_set4_filtered <- lenient_set4 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

# * * * * * Those preferentially translated in protrusions 

## enhanced_translation_in_protrusion >= 3
lenient_set5_target <- df_annotated %>%
  filter(enhanced_translation_in_protrusion >= 3) %>%
  pull(gene_id)

## How many genes in this set? = 210
lenient_set5_target %>% length()

## runTopGO
lenient_set5 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = lenient_set5_target)
}, .id = "ontology")

lenient_set5_filtered <- lenient_set5 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)

# * * * * * Those preferentially translated in protrusions & enriched 

## enhanced_translation_in_protrusion >= 3 & enriched_in_protrusion >= 3
lenient_set6_target <- df_annotated %>%
  filter(enhanced_translation_in_protrusion >= 3 & enriched_in_protrusion >= 3) %>%
  pull(gene_id)

## How many genes in this set? = 26
lenient_set6_target %>% length()

## runTopGO
lenient_set6 <- map_dfr(ontologies, function(ontology){
  runTopGO(ontology = ontology,
           species = "mouse",
           backgroundGenes = Mmus_brain_glia_txome,
           targetedGenes = lenient_set6_target)
}, .id = "ontology")

lenient_set6_filtered <- lenient_set6 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(Significant > 10) %>%
  arrange(bonferroni)


```


## Plot Gene Ontology Summary

```{r}
## Environment
library(tidyverse)
library(hrbrthemes)
library(patchwork)

```


```{r}
## Plot helpers
ontology_order <- c("BP", "MF", "CC")
ontology_colours <- c("#d1ab75", "#8fd175", "#d175b8")

## Plots
lenient_set1 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(case_when(
    ontology == "CC" ~ Significant > 200,
    TRUE ~ Significant > 150
  )) %>%
  mutate(plot_FDR = -log10(bonferroni)) %>%
  mutate(ontology = fct_relevel(ontology, ontology_order)) %>%
  ggplot(aes(x = reorder(Term, plot_FDR), y = plot_FDR)) +
  geom_segment(aes(xend = reorder(Term, plot_FDR), y = 0, yend = plot_FDR), colour = "gray70") +
  geom_point(aes(colour = ontology), size = 4) +
  geom_label(aes(y = plot_FDR + 2, label = Significant), cex = 2) + 
  scale_colour_manual(values = ontology_colours) +
  labs(title = "RNA present in glial protrusions (n = 2,755)",
       y = "-log10 Bonferroni p-value",
       colour = "Gene Ontology") + 
  coord_flip() +
  theme_classic() + 
  theme(axis.title.y = element_blank(),
        legend.position = "right") -> go_plot1

lenient_set2 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 2) %>%
  filter(case_when(
    ontology == "CC" ~ Significant > 35,
    TRUE ~ Significant > 25
  )) %>%
  mutate(plot_FDR = -log10(bonferroni)) %>%
  mutate(ontology = fct_relevel(ontology, ontology_order)) %>%
  ggplot(aes(x = reorder(Term, plot_FDR), y = plot_FDR)) +
  geom_segment(aes(xend = reorder(Term, plot_FDR), y = 0, yend = plot_FDR), colour = "gray70") +
  geom_point(aes(colour = ontology), size = 4) +
  geom_label(aes(y = plot_FDR + 1, label = Significant), cex = 2) + 
  scale_colour_manual(values = ontology_colours) +
  labs(title = "RNA enriched in glial protrusions (n = 491)",
       y = "-log10 Bonferroni p-value",
       colour = "Gene Ontology") + 
  coord_flip() +
  theme_classic() + 
  theme(axis.title.y = element_blank(),
        legend.position = "right") -> go_plot2

lenient_set5 %>%
  filter(bonferroni < 0.05) %>%
  filter(foldEnrichment > 1.75) %>%
  filter(case_when(
    ontology == "CC" ~ Significant > 10,
    TRUE ~ Significant > 10
  )) %>%
  mutate(plot_FDR = -log10(bonferroni)) %>%
  mutate(ontology = fct_relevel(ontology, ontology_order)) %>%
  ggplot(aes(x = reorder(Term, plot_FDR), y = plot_FDR)) +
  geom_segment(aes(xend = reorder(Term, plot_FDR), y = 0, yend = plot_FDR), colour = "gray70") +
  geom_point(aes(colour = ontology), size = 4) +
  geom_label(aes(y = plot_FDR + 0.5, label = Significant), cex = 2) + 
  scale_colour_manual(values = ontology_colours) +
  labs(title = "RNA preferentially translated in glial protrusions (n = 210)",
       y = "-log10 Bonferroni p-value",
       colour = "Gene Ontology") + 
  coord_flip() +
  theme_classic() + 
  theme(axis.title.y = element_blank(),
        legend.position = "right") -> go_plot3

## Patchwork
(go_plot1 + go_plot2) / (plot_spacer() + go_plot3) + plot_layout(guides = "collect")

ggsave("./RNAseq/plots/GO_patchwork.pdf", 
       height = 10, width = 16)

# save.image("./RNAseq/post_analysis_RData/GO_analysis.RData")

```


# PCA ANALYSIS

```{r}
## Environment
library(tidyverse)
library(biomaRt)
library(DESeq2)
library(pheatmap)
library(RColorBrewer)
library(hrbrthemes)

```

## Get count matrix

```{r}
## Get count matrix of Mmus kallisto datasets
Mmus_kallisto <- readRDS("./RNAseq/quant_results/Mmus_kallisto_tximport_gni.RDS")
Mmus_counts <- Mmus_kallisto$counts %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  as_tibble() %>%
  mutate(gene_id = str_replace(gene_id, "\\..*", "")) %>%
  dplyr::rename_with(.fn = ~ str_replace(.x, "Mmus_", ""),
                     .cols = starts_with("Mmus_"))

## Get count matrix of Rnor kallisto datasets
Rnor_kallisto <- readRDS("./RNAseq/quant_results/Rnor_kallisto_tximport_gni.RDS")
Rnor_counts <- Rnor_kallisto$counts %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  as_tibble() %>%
  mutate(gene_id = str_replace(gene_id, "\\..*", "")) 

## Convert Rnor genes to mouse genes
rat_mart98 <- useMart("ENSEMBL_MART_ENSEMBL", dataset = "rnorvegicus_gene_ensembl", 
                   host = "https://sep2019.archive.ensembl.org")
id_dictionary <- getBM(attributes = c("ensembl_gene_id", "mmusculus_homolog_ensembl_gene", "mmusculus_homolog_orthology_type"),
      mart = rat_mart98,
      filters = "ensembl_gene_id",
      values = Rnor_counts$gene_id) %>%
  dplyr::select(-mmusculus_homolog_orthology_type)

Rnor_counts_converted <- Rnor_counts %>%
  left_join(id_dictionary, by = c("gene_id" = "ensembl_gene_id")) %>%
  filter(str_detect(mmusculus_homolog_ensembl_gene, "ENS")) %>%
  group_by(mmusculus_homolog_ensembl_gene) %>%
  dplyr::select(-gene_id) %>%
  dplyr::select("gene_id" = mmusculus_homolog_ensembl_gene, everything()) %>%
  pivot_longer(cols = 2:last_col(), names_to = "library", values_to = "count") %>%
  group_by(gene_id, library) %>%
  summarise(counts = sum(count)) %>%
  ungroup() %>%
  pivot_wider(id_cols = gene_id, values_from = counts, names_from = library) %>%
  dplyr::rename_with(.fn = ~ str_replace(.x, "Rnor_", ""),
                     .cols = starts_with("Rnor_"))

## Combine Mmus counts with the converted Rnor counts and add microglia data
vasek_counts <- readRDS("./RNAseq/Rawdata_unavailable/Vasek2021/Vasek2021_counts_dataframe_names_converted.RDS") %>%
  dplyr::select(gene_id, contains("induced")) %>%
  dplyr::rename_with(.cols = contains("induced"),
                     .fn = ~ str_replace(.x, "_induced", ""))

counts_all <- full_join(Mmus_counts, Rnor_counts_converted, by = "gene_id") %>%
  full_join(vasek_counts, by = "gene_id") %>%
  mutate(
    across(2:last_col(), ~replace_na(.x, 0))
  ) %>%
  column_to_rownames(var = "gene_id")

```


## Perform PCA with DESeq2

### Both transcriptome and translatome

```{r}
counts_meta <- tibble(rownames = colnames(counts_all)) %>%
  mutate(compartment = case_when(
    str_detect(rownames, "protrusion") ~ "protrusion",
    str_detect(rownames, "soma") ~ "soma"
  )) %>%
  mutate(celltype = str_extract(rownames, "[^_]+")) %>%
  mutate(librarytype = case_when(
    str_detect(rownames, "txn") ~ "transcriptome",
    str_detect(rownames, "trap") ~ "translatome",
  )) %>%
  column_to_rownames(var = "rownames")

dds <- DESeqDataSetFromMatrix(countData = round(as.matrix(counts_all)),
                              colData = counts_meta,
                              design = ~ compartment + celltype + librarytype)

vsd <- vst(dds, blind=FALSE)

sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

plotPCA(vsd, intgroup = c("celltype", "librarytype", "compartment"), ntop = 1000)

```

### Separate transcriptome and translatome PCA

```{r}
## Helper function to get variance stabilised data
getVSD <- function(x){
  counts <- counts_all %>%
    dplyr::select(matches(x))
  meta <- tibble(rownames = colnames(counts)) %>%
    mutate(compartment = case_when(
      str_detect(rownames, "protrusion") ~ "protrusion",
      str_detect(rownames, "soma") ~ "soma"
    )) %>%
    mutate(celltype = str_extract(rownames, "[^_]+")) %>%
    column_to_rownames(var = "rownames")
  dds <- DESeqDataSetFromMatrix(countData = round(as.matrix(counts)),
                                colData = meta,
                                design = ~ celltype + compartment)
  vsd <- vst(dds, blind = FALSE)
  return(vsd)
}

## Transcriptome
vsd_txn <- getVSD("txn")
plotPCA(vsd_txn, intgroup = c("celltype", "compartment"), ntop = 1000)

## Translatome
vsd_trap <- getVSD("trap")
plotPCA(vsd_trap, intgroup = c("celltype", "compartment"), ntop = 1000)

## Translatome with Astrocytes only
counts <- counts_all %>%
  dplyr::select(contains("Astrocytes") & contains("trap"))
meta <- tibble(rownames = colnames(counts)) %>%
  mutate(compartment = case_when(
    str_detect(rownames, "protrusion") ~ "protrusion",
    str_detect(rownames, "soma") ~ "soma"
  )) %>%
  mutate(celltype = str_extract(rownames, "[^_]+")) %>%
  column_to_rownames(var = "rownames")
dds <- DESeqDataSetFromMatrix(countData = round(as.matrix(counts)),
                              colData = meta,
                              design = ~ celltype + compartment)
vsd_astrocyte_trap <- vst(dds, blind = FALSE)
plotPCA(vsd_astrocyte_trap, intgroup = c("celltype", "compartment"), ntop = 50000)
plotPCA(vsd_astrocyte_trap, intgroup = c("celltype", "compartment"), ntop = 50000, 
        returnData = TRUE) -> PCA_df

## Make a pretty PCA plot
PCA_df %>%
  mutate(compartment = case_when(
    compartment == "protrusion" ~ "Protrusion",
    compartment == "soma" ~ "Soma"
  )) %>%
  ggplot(aes(x = PC1, y = PC2, colour = celltype, alpha = compartment)) +
  geom_point(size = 4, stroke = 0) +
  scale_alpha_manual(values = c(0.9, 0.4)) +
  scale_colour_ipsum() + 
  labs(title = "Principal component analysis",
       subtitle = "Astrocyte TRAP-seq libraries",
       x = "PC1: 31% Variance",
       y = "PC2: 23% Variance",
       colour = "Cell type",
       alpha = "Compartment") +
  theme_classic()

ggsave("./RNAseq/plots/PCA_Astrocytes_trap.pdf", 
       width = 6,
       height = 5)

```

# SFARI OVERLAP

```{r}

```


# CHEKULAEVA OVERLAP

```{r}
library(tidyverse)
library(pathwork)
library(readxl)
library(janitor)
library(ggVennDiagram)

```

```{r}
## Get glial genes
df_annotated <- readRDS("./RNAseq/summary_table/df_annotated.RDS")

glia_protrusion_enriched_genes <- df_annotated %>%
  filter(enriched_in_protrusion >= 3) 
glia_protrusion_enhanced_translation <- df_annotated %>%
  filter(enhanced_translation_in_protrusion >= 3)

## Get Chekulaeva neuronal data 
neurite_enriched <- read_excel("./RNAseq/external_data/Supplementary online tables_neurite-enriched.xlsx",
                           skip = 1) %>%
  clean_names()

neurite_top <- neurite_enriched %>%
  filter(datasets_with_significant_neurite_enrichment_p_0_1 >= 6)
  # slice_max(datasets_with_significant_neurite_enrichment_p_0_1, n = 500)

neurite_translation <- neurite_enriched %>%
  filter(studies_with_neurite_ribosome_association >= 2)

# * * * * * Glial protrusion enriched RNA

enriched_RNA_glia_neurite_id <- intersect(glia_protrusion_enriched_genes$gene_id, neurite_top$gene_id)

enriched_RNA_venn_df <- list(Glia_protrusion_over_3datasets = glia_protrusion_enriched_genes$gene_id, 
                             Neurite_enrichement_over_6datasets = neurite_top$gene_id)

ggVennDiagram(enriched_RNA_venn_df)

enriched_RNA_glia_neurite <- df_annotated %>%
  filter(gene_id %in% enriched_RNA_glia_neurite_id)

# * * * * * Glial protrusion enhanced translation

enhanced_translation_glia_neurite_id <- intersect(glia_protrusion_enhanced_translation$gene_id, neurite_translation$gene_id)

enhanced_translation_venn_df <- list(Glia_translation_over_3datasets = glia_protrusion_enhanced_translation$gene_id, 
                                     Neurite_translation_over_2datasets = neurite_translation$gene_id)

ggVennDiagram(enhanced_translation_venn_df)

enhanced_translation_glia_neurite <- df_annotated %>%
  filter(gene_id %in% enhanced_translation_glia_neurite_id)

```


























