---
title: "RNAseq_analysis"
author: "Jeff Lee"
date: "16/12/2021"
output: html_document
---

# INTRODUCTION 

R Markdown document to re-analyse RNA-seq from glial end-processes. 
Raw data handling and kallisto mappings are performed in @mprocessor1 while DESeq2 is run locally. 

# RNA SEQUENCING ANALYSIS

## Raw Data Handling

### Download fastq data from SRA

Fastq curl links are obtained from https://sra-explorer.info

```{bash}
## @mprocessor1
## project dir = /usr/people/sjoh4548/Data/Sequencing

mkdir -p DGS_review/Rawfastq

## Get curl links
cd DGS_review/Rawfastq

## * * * * * Naming convention
## species_celltype_compartment_txn/trap + repeat number_(read number if PE).fastq.gz
## e.g. Rnor_Oligodendrocytes_soma_txn1_read1.fastq.gz

### Azevedo 2018 (Rnor) - the only PE sequenced data 
# physical separation of OPC soma and membrane protrusions
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR628/009/SRR6288019/SRR6288019_1.fastq.gz -o Rnor_Oligodendrocytes_soma_txn1_read1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR628/009/SRR6288019/SRR6288019_2.fastq.gz -o Rnor_Oligodendrocytes_soma_txn1_read2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR628/000/SRR6288020/SRR6288020_1.fastq.gz -o Rnor_Oligodendrocytes_soma_txn2_read1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR628/000/SRR6288020/SRR6288020_2.fastq.gz -o Rnor_Oligodendrocytes_soma_txn2_read2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR628/002/SRR6288022/SRR6288022_1.fastq.gz -o Rnor_Oligodendrocytes_protrusion_txn2_read1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR628/002/SRR6288022/SRR6288022_2.fastq.gz -o Rnor_Oligodendrocytes_protrusion_txn2_read2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR628/001/SRR6288021/SRR6288021_1.fastq.gz -o Rnor_Oligodendrocytes_protrusion_txn1_read1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR628/001/SRR6288021/SRR6288021_2.fastq.gz -o Rnor_Oligodendrocytes_protrusion_txn1_read2.fastq.gz

### Sakers 2017 (Mmus)
# Cortex transcript, cortex trap, SN transcript, SN trap
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/009/SRR2862549/SRR2862549.fastq.gz -o Mmus_Astrocytes-A_wholecell_trap1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/006/SRR2862546/SRR2862546.fastq.gz -o Mmus_Astrocytes-A_wholecell_txn1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/001/SRR2862551/SRR2862551.fastq.gz -o 
Mmus_Astrocytes-A_wholecell_trap3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/000/SRR2862550/SRR2862550.fastq.gz -o Mmus_Astrocytes-A_wholecell_trap2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/007/SRR2862547/SRR2862547.fastq.gz -o 
Mmus_Astrocytes-A_wholecell_txn2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/008/SRR2862548/SRR2862548.fastq.gz -o Mmus_Astrocytes-A_wholecell_txn3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/002/SRR2862552/SRR2862552.fastq.gz -o Mmus_Astrocytes-A_protrusion_txn1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/003/SRR2862553/SRR2862553.fastq.gz -o Mmus_Astrocytes-A_protrusion_txn2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/004/SRR2862554/SRR2862554.fastq.gz -o Mmus_Astrocytes-A_protrusion_txn3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/005/SRR2862555/SRR2862555.fastq.gz -o Mmus_Astrocytes-A_protrusion_trap1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/006/SRR2862556/SRR2862556.fastq.gz -o Mmus_Astrocytes-A_protrusion_trap2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR286/007/SRR2862557/SRR2862557.fastq.gz -o Mmus_Astrocytes-A_protrusion_trap3.fastq.gz

### Boulay 2017 (Mmus)

#### Translatome
# translatome of whole astrocytes extracted following the bacTRAP protocol was compared to a translatome of astrocyte perivascular endfeet
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/009/SRR2352159/SRR2352159.fastq.gz -o 
Mmus_Astrocytes-B_translatome_Endfeet2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/006/SRR2352156/SRR2352156.fastq.gz -o 
Mmus_Astrocytes-B_translatome_wholecell2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/007/SRR2352157/SRR2352157.fastq.gz -o 
Mmus_Astrocytes-B_translatome_wholecell3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/005/SRR2352155/SRR2352155.fastq.gz -o 
Mmus_Astrocytes-B_translatome_wholecell1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/008/SRR2352158/SRR2352158.fastq.gz -o 
Mmus_Astrocytes-B_translatome_Endfeet1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/000/SRR2352160/SRR2352160.fastq.gz -o 
Mmus_Astrocytes-B_translatome_Endfeet3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/002/SRR2352162/SRR2352162.fastq.gz -o 
Mmus_Astrocytes-B_translatome_Endfeet5.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/001/SRR2352161/SRR2352161.fastq.gz -o 
Mmus_Astrocytes-B_translatome_Endfeet4.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR235/003/SRR2352163/SRR2352163.fastq.gz -o 
Mmus_Astrocytes-B_translatome_Endfeet6.fastq.gz


### Castro 2020 (Mmus)
# They analyse FACS sortes transcriptomes of 3 cell types, and Perisynaptic Schwann Cells are double labelled with S100Î²-GFP+;NG2-dsRed+ 
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/023/SRR12045623/SRR12045623.fastq.gz -o 
Mmus_PCSc_1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/021/SRR12045621/SRR12045621.fastq.gz -o 
Mmus_PCSc_2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/027/SRR12045627/SRR12045627.fastq.gz -o 
Mmus_PCSc_3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/029/SRR12045629/SRR12045629.fastq.gz -o 
Mmus_PCSc_4.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/035/SRR12045635/SRR12045635.fastq.gz -o 
Mmus_PCSc_5.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/037/SRR12045637/SRR12045637.fastq.gz -o 
Mmus_PCSc_6.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/042/SRR12045642/SRR12045642.fastq.gz -o 
Mmus_PCSc_7.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/040/SRR12045640/SRR12045640.fastq.gz -o 
Mmus_PCSc_8.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/048/SRR12045648/SRR12045648.fastq.gz -o 
Mmus_PCSc_9.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/050/SRR12045650/SRR12045650.fastq.gz -o 
Mmus_PCSc_10.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/055/SRR12045655/SRR12045655.fastq.gz -o 
Mmus_PCSc_11.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR120/053/SRR12045653/SRR12045653.fastq.gz -o 
Mmus_PCSc_12.fastq.gz

### Mazare 2020 (Mmus)
# This is TRAP ONLY: RNA-seq raw data of ribosome-bound mRNAs extracted from whole astrocytes and PAPs by purifying GFP ribosomes and sequencing RNAs attached to them
# Experiment was done 3 times, each time with multiple technical replicates
# Perisynaptic Astrocytic Processes (PAPs) versus full astrocytes polysomal transcriptome from dorsal hippocampus in Aldh1l1:L10a-eGFP mice
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/094/SRR10877094/SRR10877094.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap1_tech6.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/090/SRR10877090/SRR10877090.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap1_tech2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/089/SRR10877089/SRR10877089.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap1_tech1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/091/SRR10877091/SRR10877091.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap1_tech3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/092/SRR10877092/SRR10877092.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap1_tech4.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/093/SRR10877093/SRR10877093.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap1_tech5.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/095/SRR10877095/SRR10877095.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap1_tech7.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/097/SRR10877097/SRR10877097.fastq.gz -o Mmus_Astrocytes-C_PAP_trap1_tech1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/096/SRR10877096/SRR10877096.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap1_tech8.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/098/SRR10877098/SRR10877098.fastq.gz -o Mmus_Astrocytes-C_PAP_trap1_tech2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/099/SRR10877099/SRR10877099.fastq.gz -o Mmus_Astrocytes-C_PAP_trap1_tech3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/000/SRR10877100/SRR10877100.fastq.gz -o Mmus_Astrocytes-C_PAP_trap1_tech4.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/004/SRR10877104/SRR10877104.fastq.gz -o Mmus_Astrocytes-C_PAP_trap1_tech8.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/002/SRR10877102/SRR10877102.fastq.gz -o Mmus_Astrocytes-C_PAP_trap1_tech6.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/003/SRR10877103/SRR10877103.fastq.gz -o Mmus_Astrocytes-C_PAP_trap1_tech7.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/001/SRR10877101/SRR10877101.fastq.gz -o Mmus_Astrocytes-C_PAP_trap1_tech5.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/005/SRR10877105/SRR10877105.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap2_tech1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/006/SRR10877106/SRR10877106.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap2_tech2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/011/SRR10877111/SRR10877111.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap2_tech7.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/010/SRR10877110/SRR10877110.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap2_tech6.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/007/SRR10877107/SRR10877107.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap2_tech3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/008/SRR10877108/SRR10877108.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap2_tech4.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/009/SRR10877109/SRR10877109.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap2_tech5.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/012/SRR10877112/SRR10877112.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap2_tech8.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/015/SRR10877115/SRR10877115.fastq.gz -o Mmus_Astrocytes-C_PAP_trap2_tech3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/014/SRR10877114/SRR10877114.fastq.gz -o Mmus_Astrocytes-C_PAP_trap2_tech2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/013/SRR10877113/SRR10877113.fastq.gz -o Mmus_Astrocytes-C_PAP_trap2_tech1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/016/SRR10877116/SRR10877116.fastq.gz -o Mmus_Astrocytes-C_PAP_trap2_tech4.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/017/SRR10877117/SRR10877117.fastq.gz -o Mmus_Astrocytes-C_PAP_trap2_tech5.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/018/SRR10877118/SRR10877118.fastq.gz -o Mmus_Astrocytes-C_PAP_trap2_tech6.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/019/SRR10877119/SRR10877119.fastq.gz -o Mmus_Astrocytes-C_PAP_trap2_tech7.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/021/SRR10877121/SRR10877121.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap3_tech1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/022/SRR10877122/SRR10877122.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap3_tech2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/020/SRR10877120/SRR10877120.fastq.gz -o Mmus_Astrocytes-C_PAP_trap2_tech8.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/023/SRR10877123/SRR10877123.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap3_tech3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/024/SRR10877124/SRR10877124.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap3_tech4.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/026/SRR10877126/SRR10877126.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap3_tech6.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/025/SRR10877125/SRR10877125.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap3_tech5.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/027/SRR10877127/SRR10877127.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap3_tech7.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/029/SRR10877129/SRR10877129.fastq.gz -o Mmus_Astrocytes-C_PAP_trap3_tech1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/028/SRR10877128/SRR10877128.fastq.gz -o Mmus_Astrocytes-C_wholecell_trap3_tech8.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/030/SRR10877130/SRR10877130.fastq.gz -o Mmus_Astrocytes-C_PAP_trap3_tech2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/032/SRR10877132/SRR10877132.fastq.gz -o Mmus_Astrocytes-C_PAP_trap3_tech4.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/034/SRR10877134/SRR10877134.fastq.gz -o Mmus_Astrocytes-C_PAP_trap3_tech6.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/031/SRR10877131/SRR10877131.fastq.gz -o Mmus_Astrocytes-C_PAP_trap3_tech3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/036/SRR10877136/SRR10877136.fastq.gz -o Mmus_Astrocytes-C_PAP_trap3_tech8.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/033/SRR10877133/SRR10877133.fastq.gz -o Mmus_Astrocytes-C_PAP_trap3_tech5.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR108/035/SRR10877135/SRR10877135.fastq.gz -o Mmus_Astrocytes-C_PAP_trap3_tech7.fastq.gz

### Thakurela 2016 (Mmus)
# This is myelin microdissection at various stages of mouse development
# Idk if we should compare all or choose oldest/6 or 24mo
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/006/SRR3181086/SRR3181086.fastq.gz -o 
Mmus_Myelin_P75_R2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/003/SRR3181083/SRR3181083.fastq.gz -o 
Mmus_Myelin_P18_R2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/007/SRR3181087/SRR3181087.fastq.gz -o 
Mmus_Myelin_P75_R3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/002/SRR3181082/SRR3181082.fastq.gz -o 
Mmus_Myelin_P18_R1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/005/SRR3181085/SRR3181085.fastq.gz -o 
Mmus_Myelin_P75_R1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/004/SRR3181084/SRR3181084.fastq.gz -o 
Mmus_Myelin_P18_R3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/008/SRR3181088/SRR3181088.fastq.gz -o 
Mmus_Myelin_6mo_R1.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/000/SRR3181090/SRR3181090.fastq.gz -o 
Mmus_Myelin_6mo_R3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/009/SRR3181089/SRR3181089.fastq.gz -o 
Mmus_Myelin_6mo_R2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/003/SRR3181093/SRR3181093.fastq.gz -o 
Mmus_Myelin_24mo_R3.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/002/SRR3181092/SRR3181092.fastq.gz -o 
Mmus_Myelin_24mo_R2.fastq.gz
curl -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR318/001/SRR3181091/SRR3181091.fastq.gz -o 
Mmus_Myelin_24mo_R1.fastq.gz

### Vasek 2021 (Mmus)




cd ../..

```

### Run fastQC (optional)

```{bash}
mkdir -p DSG_review/Rawfastq/FastQC

# Run fastqc
for file in DSG_review/Rawfastq/*.fastq.gz
do
echo "Processing file $file"
fastqc $file -o DSG_review/Rawfastq/FastQC
echo ""
done 

# Run multiqc
mkdir -p DSG_review/Rawfastq/multiqc

multiqc -f \
DSG_review/Rawfastq/FastQC/*_fastqc.zip \
-o DSG_review/Rawfastq/multiqc \
-n MultiQC_FastQC_Sequencing-runs_raw_report

```

## Remove Adapters

* Illumina Tru-seq AGATCGGAAGAGCACACGTCTGAACTCCAGTCA
* NuGen SPIA CTTTGTGTTTGA

```{bash}
mkdir -p DSG_review/trimmed

for file in DSG_review/Rawfastq/*.fastq.gz
do
echo "Processing file $file"
base=$(basename "$file") 
outfile="$(echo ${base} |sed -e 's/.fastq.gz/_trimmed.fastq.gz/')" 
cutadapt \
-a AGATCGGAAGAGCACACGTCTGAACTCCAGTCA \
-a CTTTGTGTTTGA \
--times 2 \
--cores=4 \
-m 30 \
--trim-n \
--report=minimal \
-o DSG_review/trimmed/$outfile \
$file
echo ""
done

```

## Filter Contaminant Reads

* Ribosomal RNA
* Mitochondrial ribosomal RNA
* PolyA/C ...
* PhiX

```{bash}
mkdir -p DSG_review/filtered

for file in DSG_review/trimmed/*_trimmed.fastq.gz
do
echo "Processing file $file"
base=$(basename "$file") 
outfile="$(echo ${base} |sed -e 's/_trimmed.fastq.gz/_filtered.fastq.gz/')" 
outstat="$(echo ${base} |sed -e 's/_trimmed.fastq.gz/_filtered.stat.txt/')" 
bbduk.sh in=$file \
out=DSG_review/filtered/$outfile \
ref=GENOMEDIR/contaminants/Dmel_lrRNA_large.fasta,GENOMEDIR/contaminants/Dmel_rRNA_NCBI.fasta,GENOMEDIR/contaminants/polyA_polyC.fasta,GENOMEDIR/contaminants/phix174_ill.ref.fa.gz \
k=27 \
hdist=1 \
-Xmx32g \
stats=DSG_review/filtered/$outstat
echo ""
done

```

## Mapping With Kallisto

### Download genomes and annotations

* ENSEMBL mouse 96
* ENSEMBL rat 98

```{bash}
## Mouse 
mkdir -p GENOMEDIR/Mmus
cd GENOMEDIR/Mmus
wget http://ftp.ensembl.org/pub/release-96/fasta/mus_musculus/dna/Mus_musculus.GRCm38.dna.toplevel.fa.gz 
wget http://ftp.ensembl.org/pub/release-96/gtf/mus_musculus/Mus_musculus.GRCm38.96.gtf.gz
cd .. 

## Rat
mkdir -p GENOMEDIR/Rnor
cd GENOMEDIR/Rnor
wget http://ftp.ensembl.org/pub/release-98/fasta/rattus_norvegicus/dna/Rattus_norvegicus.Rnor_6.0.dna.toplevel.fa.gz
wget http://ftp.ensembl.org/pub/release-98/gtf/rattus_norvegicus/Rattus_norvegicus.Rnor_6.0.98.gtf.gz
cd .. 

```


### Create kallisto index with introns

```{bash}
## Mouse
mkdir -p GENOMEDIR/Mmus/kallisto

kb ref -i GENOMEDIR/Mmus/kallisto/kallisto_index_Mmus_ensembl96_cdna_intron \
-g GENOMEDIR/Mmus/kallisto/kb_output_t2g.txt \
-f1 GENOMEDIR/Mmus/kallisto/kb_output_cdna.fa \
-f2 GENOMEDIR/Mmus/kallisto/kb_output_intron.fa \
-c1 GENOMEDIR/Mmus/kallisto/kb_output_cdna_t2c.txt \
-c2 GENOMEDIR/Mmus/kallisto/kb_output_intron_t2c.txt \
--workflow lamanno \
GENOMEDIR/Mmus/Mus_musculus.GRCm38.dna.toplevel.fa.gz \
GENOMEDIR/Mmus/Mus_musculus.GRCm38.96.gtf.gz

## Rat
mkdir -p GENOMEDIR/Rnor/kallisto

kb ref -i GENOMEDIR/Rnor/kallisto/kallisto_index_Rnor_ensembl98_cdna_intron \
-g GENOMEDIR/Rnor/kallisto/kb_output_t2g.txt \
-f1 GENOMEDIR/Rnor/kallisto/kb_output_cdna.fa \
-f2 GENOMEDIR/Rnor/kallisto/kb_output_intron.fa \
-c1 GENOMEDIR/Rnor/kallisto/kb_output_cdna_t2c.txt \
-c2 GENOMEDIR/Rnor/kallisto/kb_output_intron_t2c.txt \
--workflow lamanno \
GENOMEDIR/Rnor/Rattus_norvegicus.Rnor_6.0.dna.toplevel.fa.gz \
GENOMEDIR/Rnor/Rattus_norvegicus.Rnor_6.0.98.gtf.gz


```

### Pseudoalign to transcriptome

```{bash}
mkdir -p DSG_review/kallisto_results

for file in DSG_review/filtered/*.fastq.gz 
do 
echo "Processing file $file" 
base=$(basename "$file") 
outfolder="$(echo ${base} |sed -e 's/_filtered.fastq.gz//')" 
mkdir -p DSG_review/kallisto_results/$outfolder

kallisto quant \
-i GENOMEDIR/kallisto/kallisto_index_dmel_ensembl99_cdna_intron \
-o DSG_review/kallisto_results/$outfolder \
-b 100 \
--single \
-l 200 \
-s 30 \
-t 4 \
$file 

echo ""
done

```

## Preprocessing Data Tables

### Get TPM dataframe

```{r}
library(tidyverse)

```

### Get count data for DESeq2 

```{r}
library(tidyverse)
library(tximport)
library(rhdf5)

## Get sample metadata
sample_id <- dir("./data/external/DGS_review/kallisto_results/")
kal_dirs <- file.path("./data/external/DGS_review/kallisto_results", sample_id)

files <- file.path(kal_dirs, "abundance.h5")
names(files) <- sample_id # names for the matrices
all(file.exists(files))

## Get tx2gene table 
tx2gene <- read_csv("./data/external/Dmel_tx2gene_ENSEMBL_v99_cdna_intron.csv")
t2g <- tx2gene %>%
  dplyr::select(tx_id, gene_id)

## Import with kallisto
gni_kallisto <- tximport(files,
                         type = "kallisto",
                         tx2gene = t2g)

head(gni_kallisto$counts)
head(gni_kallisto$length)

dir.create("./data/external/DGS_review/DGE")
saveRDS(gni_kallisto, "./data/external/DGS_review/DGE/gni_kallisto_full.RDS")

```








