---
title: "Upset-from-df"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Environment

```{r}
library(tidyverse)
library(UpSetR)
library(ggplot2)
```

## Import data and clean-up

```{r}
## * * * * * Define a parsing function
combine_glial_list <- function(file){
  # get filename that will be added as column name
  filename  <- basename(file)
  # get tibble, remove duplicates, and create a column to indicate category "1"
  gene_list <- read_csv(file) %>%
    distinct() %>%
    mutate(!!filename := 1) 
}

## * * * * * Choose directory and list .csv files
list_dir   <- "./Data"
list_files <- list.files(list_dir, pattern = ".csv", full.names = TRUE)

## * * * * * Run combine_glial_list function over list_files 
gene_list_df <- map(list_files, combine_glial_list) %>%
  # full_join dataframes parsed by the function
  reduce(full_join, by = "Human_Symbol") %>%
  # remove ".csv" extension from column names 
  rename_all(~str_remove(., ".csv")) %>%
  # replace NA with 0 
  mutate_all(replace_na, replace = 0) %>%
  # column for number of intersects
  mutate(n_intersect = rowSums(across(where(is.numeric)))) %>%
  dplyr::select(Human_Symbol, n_intersect, everything())

write.csv(gene_list_df, "gene_list_df.csv")

# saveRDS(gene_list_df, "./RDS/gene_list_df.RDS")

```

## Create upset plot

```{r}
## * * * * * UpSet plot
# Re-organise dataframe for upset function (only takes dataframe not tibble!)
upset_df <- gene_list_df %>%
  dplyr::select(-n_intersect) %>%
  mutate_if(is.numeric, as.integer) %>%
  as.data.frame()

# Plot
upset(upset_df,
      nsets = 10,
      nintersects = 23,
      mainbar.y.label = "Intersecting transcripts",
      sets.x.label    = "Number of transcripts per set",
      point.size = 5, 
      line.size = 1, 
      mb.ratio = c(0.6, 0.4),
      order.by = c("freq"),
      scale.sets = "identity", 
      text.scale = 2,
      set_size.show = TRUE, 
      set_size.numbers_size = 8,
      set_size.scale_max = 1500)

#ggsave("upset_glial.png", last_plot(), dpi = 300)

```

## Top 20 genes with highest number of intersections

```{r}
topNeuronGenes <- gene_list_df %>%
#  filter(Neuron == 0) %>%
  arrange(desc(n_intersect)) %>% head(328)
write.csv(topGenes, "topNeuronGenes.csv")
```

## Genes that intersect between Azevedo and Castro datasets

```{r}
gene_list_df %>%
  filter(Azevedo == 1 & Castro == 1) %>%
  arrange(Human_Symbol) %>% pull(Human_Symbol)

```

## Which genes were found in only one study?

```{r}
one_study_genes <- gene_list_df %>%
  filter(n_intersect == 1)
```

## Which genes intersect with neuronal genes?

```{r}
neuronal_intersect<- gene_list_df %>%
  filter(Neuron == 1) %>%
  filter(n_intersect >= 2)
neuronal_intersect %>% nrow()

# How many genes in von_Kuglegen? Check
gene_list_df %>%
  dplyr::select(von_Kuglegen) %>%
  filter(von_Kuglegen == 1) %>%
  summarise(sum = n())

tt <- read_csv("./Data/Microglia.csv") 

tt %>% distinct() %>% nrow()

```

## How many neural genes in the core transcriptome?

```{r}
topGenes <- read_csv("./withoutNeuro.csv")
neuronal_only<- gene_list_df %>%
  filter(Neuron == 1)

neural_glial <- inner_join(neuronal_only, topGenes, by = "Human_Symbol")
write.csv(neural_glial, "neural_glial_overlap.csv")
```

## % Overlap with k
```{r}
how_many_overlap <- gene_list_df %>%
  filter(Neuron == 1) %>%
  dplyr::select(Microglia) %>% colSums()

how_many_in_set <- gene_list_df %>%
  filter(Microglia == 1) %>% nrow()

p_overlap <- (how_many_overlap/how_many_in_set) * 100

```
## Plot % overlaps
```{r}
p_overlap_frame <- read_csv("./percentages_summaries_all.csv") %>%
  mutate(Dataset = fct_relevel(Dataset, "Endfeet", "PSCs","Oligodendrocytes", "Radial", "Astrocytes_S", "Astrocytes_T", "Astrocytes_M", "Myelin"))

# Basic barplot
p_overlap_frame %>%
  ggplot() +
  geom_col(aes(x=Dataset, y=overlap, fill = Dataset)) +
  geom_label(aes(x = Dataset, y = overlap + 1, label = Number_of_genes)) +
  labs(title = "Overlap with neuronal dataset",
       x = "Dataset",
       y = "Percentage overlap") +
  scale_fill_brewer(palette = "PuOr") +
  theme_gray(base_size = 13) + 
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "none")

# Horizontal bar plot
# p + coord_flip()
```

